freyja/read_analysis_tools.py

```diff
@@ -388,6 +400,360 @@ def filter(query_mutations, input_bam, min_site, max_site, output, refname):
     samfile.close()
     outfile.close()
 
-    print(f'Output saved to {output}')
+    print(f'filter: Output saved to {output}')
 
     return final_reads
+
+
+def covariants(input_bam, min_site, max_site, output, refname,
+               ref_fasta, gff_file, min_quality, min_count, spans_region):
+    def read_pair_generator(bam):
+        is_paired = {}
+        for read in bam.fetch(refname, min_site, max_site+1):
+            qname = read.query_name[:-1]
+            if qname not in is_paired:
+                is_paired[qname] = False
+            else:
+                is_paired[qname] = True
+
+        read_dict = {}
+        for read in bam.fetch(refname, min_site, max_site+1):
+            if not is_paired[read.query_name[:-1]]:
+                yield read, None
+                continue
+            qname = read.query_name[:-1]
+
+            if qname not in read_dict:
+                read_dict[qname] = read
+            else:
+                yield read_dict[qname], read
+                del read_dict[qname]
+
+    def get_gene(locus):
+        for gene in gene_positions:
+            start, end = gene_positions[gene]
+            if locus in range(start, end+1):
+                return gene, start
+
+    # Load gene annotations
+    gene_positions = {}
+    with open(gff_file) as f:
+        for line in f.readlines():
+            line = line.split('\t')
+            if 'gene' in line:
+                attrs = line[-1].split(';')
+                for attr in attrs:
+                    if attr.startswith('gene='):
+                        gene_name = attr.split('=')[1]
+
+                        gene_positions[gene_name] = (int(line[3]),
+                                                     int(line[4]))
+
+        # Split ORF1ab for SARS-CoV-2
+        if refname == 'NC_045512.2' and 'ORF1ab' in gene_positions:
```

Comment:
may need to be a bit more general here. There's another name for Hu-1, for example: https://www.ncbi.nlm.nih.gov/nuccore/MN908947.